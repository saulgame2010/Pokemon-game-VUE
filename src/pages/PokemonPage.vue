<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>¿Quién es este pokemon?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions v-if="showOptions" :pokemons="pokemonArr" @selection="checkAnswer" />
    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">Nuevo juego</button>
    </template>
  </div>
</template>

<script>
import PokemonOptions from "@/components/PokemonOptions.vue";
import PokemonPicture from "@/components/PokemonPicture.vue";
import getPokemonOptions from "@/helpers/getPokemonOptions";

export default {
  components: {
    PokemonOptions,
    PokemonPicture,
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      showOptions: true,
      message: "",
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },
    checkAnswer(pokemonId) {
      this.showPokemon = true;
      this.showAnswer = true;
      this.showOptions = false;
      if (pokemonId === this.pokemon.id) {
        this.message = `Adivinaste!!! Es ${this.pokemon.name}`;
      } else {
        this.message = `Fallaste!!! Es ${this.pokemon.name}`;
      }
    },
    newGame() {
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArr = [];
      this.pokemon = null;
      this.mixPokemonArray();
      this.showOptions = true;
    }
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style>
</style>