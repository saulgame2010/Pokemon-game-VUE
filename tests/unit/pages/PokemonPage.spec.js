import { shallowMount, mount } from "@vue/test-utils"
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from "../mocks/pokemons.mock";

describe('PokemonPage Component', () => {

    let wrapper;

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage);
    })

    test('Debe hacer match con el snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    })

    test('Debe de llamar el mixPokemonArr al montar', () => {
        const mixPokemonArrSpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');
        const wrapper = shallowMount(PokemonPage);
        expect(mixPokemonArrSpy).toHaveBeenCalled();
    })

    test('Debe de hacer match con el snapshot cuando cargan los pokemons', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showOptions: true,
                    message: "",
                }
            }
        });
        expect(wrapper.html()).toMatchSnapshot();
    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showOptions: true,
                    message: "",
                }
            }
        });
        const pokemonPicture = wrapper.find('pokemon-picture-stub');
        const pokemonOptions = wrapper.find('pokemon-options-stub');
        expect(pokemonPicture.exists()).toBeTruthy();
        expect(pokemonOptions.exists()).toBeTruthy();
        expect(pokemonPicture.attributes('pokemonid')).toBe('1');
        expect(pokemonOptions.attributes('pokemons')).toBeTruthy();
    })

    test('Pruebas con checkAnswer', async () => {
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    showOptions: true,
                    message: "",
                }
            }
        });
        await wrapper.vm.checkAnswer(1);
        expect(wrapper.find('h2').exists()).toBeTruthy();
        expect(wrapper.vm.showPokemon).toBeTruthy();
        expect(wrapper.find('h2').text()).toBe(`Adivinaste!!! Es ${pokemons[0].name}`);
        await wrapper.vm.checkAnswer(12);
        expect(wrapper.vm.message).toBe(`Fallaste!!! Es ${pokemons[0].name}`);
    })
})