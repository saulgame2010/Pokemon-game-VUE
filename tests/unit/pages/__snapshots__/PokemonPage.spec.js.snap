// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokemon?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage Component Debe hacer match con el snapshot 1`] = `<h1>Espere por favor...</h1>`;
